pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go

```diff
@@ -0,0 +1,131 @@
+	"io"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/tools/clusters"
+)
+
+const (
+	PluginName = "apis.kcp.dev/CRDNoOverlappingGVR"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return &crdNoOverlappingGVRAdmission{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+type crdNoOverlappingGVRAdmission struct {
+	*admission.Handler
+
+	apiBindingLister        apislisters.APIBindingLister
+	apiExportLister         apislisters.APIExportLister
+	apiResourceSchemaLister apislisters.APIResourceSchemaLister
+}
+
+// Ensure that the required admission interfaces are implemented.
+var _ = admission.ValidationInterface(&crdNoOverlappingGVRAdmission{})
+
+func (p *crdNoOverlappingGVRAdmission) SetKcpInformers(informers kcpinformers.SharedInformerFactory) {
+	p.SetReadyFunc(func() bool {
+		return informers.Apis().V1alpha1().APIBindings().Informer().HasSynced() &&
+			informers.Apis().V1alpha1().APIExports().Informer().HasSynced() &&
+			informers.Apis().V1alpha1().APIResourceSchemas().Informer().HasSynced()
+	})
+	p.apiBindingLister = informers.Apis().V1alpha1().APIBindings().Lister()
+	p.apiExportLister = informers.Apis().V1alpha1().APIExports().Lister()
+	p.apiResourceSchemaLister = informers.Apis().V1alpha1().APIResourceSchemas().Lister()
+}
+
+// Validate checks if the given CRD's Group and Resource don't overlap with other CRDs exposed via APIBindings
+func (p *crdNoOverlappingGVRAdmission) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	if a.GetResource().GroupResource() != apiextensions.Resource("customresourcedefinitions") {
+		return nil
+	}
+
+	if a.GetKind().GroupKind() != apiextensions.Kind("CustomResourceDefinition") {
+		return nil
+	}
+	crd, ok := a.GetObject().(*apiextensions.CustomResourceDefinition)
+	if !ok {
+		return fmt.Errorf("unexpected type %T", a.GetObject())
+	}
+	clusterName, err := request.ClusterNameFrom(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster from context: %w", err)
+	}
+
+	apiBindingsForCurrentClusterName, err := p.listAPIBindingsFor(clusterName)
+	if err != nil {
+		return err
+	}
+	for _, apiBindingForCurrentClusterName := range apiBindingsForCurrentClusterName {
+		apiExportClusterName, err := getAPIExportClusterName(apiBindingForCurrentClusterName)
+		if err != nil {
+			// since all exports needs to be checked we need to stop in case of an error
+			return err
+		}
+		apiExport, err := p.getAPIExport(apiExportClusterName, apiBindingForCurrentClusterName.Spec.Reference.Workspace.ExportName)
```

Comment:
I think that `CRDs` from an APIExport are the ones that are eventually created in the "current" cluster.

Reply:
try to not depend on the APIExport. The APIBinding itself (the status) must be enough for this decision.

Background: APIExport might live on a different shard and we should not break if the shard is unavailable (network partition). If something is missing for the logic in the APIBinding status, we should add it.

Reply:
oh, ic, does it mean we should reject the request if APIBinding is still progressing?

Reply:
The status of the APIBinding is what counts. This admission plugin also has to stop binding to succeed (with some good condition message).